IntegrationTests/UITests/UITest.cs

```diff
@@ -1,58 +1,177 @@
+                var form = Application.OpenForms.OfType<T>().Single();
+                if (!form.Visible)
+                {
+                    var errorMessage = $"{form.GetType().FullName} {form.Name} should be visible";
+                    form.Dispose();
+                    Assert.Fail(errorMessage);
+                }
+
+                try
+                {
+                    // Wait for potential pending asynchronous tasks triggered by the form.
+                    WaitForPendingOperations();
+                    LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test form loaded");
+
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test on main thread");
+
+                    try
+                    {
+                        await runTestAsync(form);
+                        LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test passed");
+                    }
+                    catch (Exception ex)
+                    {
+                        LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test failed", ex);
+                        throw;
+                    }
+                }
+                finally
+                {
+                    CloseOrDispose(form, $"{MethodBase.GetCurrentMethod().Name}.test");
+                }
+            });
+
+            // The following call to showForm will trigger messages.
+            // So we can be sure we don't stall waiting for idle if the application was already idle.
             Application.Idle += HandleApplicationIdle;
+            LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} subscribed");
+            try
+            {
+                try
+                {
+                    showForm();
+                    LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} form shown");
+                }
+                catch (Exception outer)
+                {
+                    try
+                    {
+                        // Wake up and join the asynchronous test operation.
+                        LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} exception",  outer);
+                        idleCompletionSource.TrySetResult(default);
+                        test.Join();
+                        LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} test joined");
+                    }
+                    catch (Exception inner)
+                    {
+                        // ignore
+                        LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} exception", inner);
+                    }
 
-            // Queue an event to make sure we don't stall if the application was already idle
-            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-            await Task.Yield();
```

Comment:
📝 This line guarantees that `WaitForIdleAsync` will not complete before `showDialog` is called.

Reply:
This can come too late because this async task does not count as an "UI operation". Hence `HandleApplicationIdle` can be called before `await Task.Yield()` is reached. Then `await idleCompletionSource.Task;` returns at once.
That's why do register for `Application.Idle` in the non-UI test thread as late as possibe, i.e. right before opening the form, which makes the UI busy.

Reply:
> Then await idleCompletionSource.Task; returns at once.

Yes, that is allowed. It means there are no pending UI operations.